src/sentry/spans/buffer_v2.py

```diff
@@ -0,0 +1,246 @@
+        redis_ttl: int = 3600,
+    ):
+        self.client: RedisCluster[bytes] | StrictRedis[bytes] = get_redis_client()
+        self.sharding_factor = sharding_factor
+        self.span_buffer_timeout = span_buffer_timeout_secs
+        self.span_buffer_root_timeout = span_buffer_root_timeout_secs
+        self.max_timeout = redis_ttl
+
+    def _is_root_span(self, span: Span) -> bool:
+        return span.parent_span_id is None or span.is_segment_span
+
+    def process_spans(self, spans: Sequence[Span], now: int):
+        queue_keys = []
+        queue_items = []
+        queue_item_has_root_span = []
+
+        is_root_span_count = 0
+        has_root_span_count = 0
+
+        with metrics.timer("sentry.spans.buffer.process_spans.step1"):
+            with self.client.pipeline(transaction=False) as p:
+                for span in spans:
+                    # (parent_span_id) -> [Span]
+                    shard = int(span.trace_id, 16) % self.sharding_factor
+                    queue_key = f"span-buf:q:{shard}"
+                    parent_span_id = span.parent_span_id or span.span_id
+
+                    is_root_span = self._is_root_span(span)
+                    if is_root_span:
+                        is_root_span_count += 1
+
+                    # hack to make redis-cluster-py pipelines work well with
+                    # scripts. we cannot use EVALSHA or "the normal way to do
+                    # scripts in sentry" easily until we get rid of
+                    # redis-cluster-py sentrywide. this probably leaves a bit of
+                    # perf on the table as we send the full lua sourcecode with every span.
+                    p.eval(
+                        add_buffer_script.script,
+                        1,
+                        f"{span.project_id}:{span.trace_id}",
+                        span.payload,
+                        "true" if is_root_span else "false",
+                        span.span_id,
+                        parent_span_id,
+                        self.max_timeout,
+                    )
+
+                    queue_keys.append(queue_key)
+
+                results = iter(p.execute())
+                for item, has_root_span in results:
+                    queue_items.append(item)
+                    queue_item_has_root_span.append(has_root_span)
+
+        with metrics.timer("sentry.spans.buffer.process_spans.step2"):
+            with self.client.pipeline(transaction=False) as p:
+                for key, item, has_root_span in zip(
+                    queue_keys, queue_items, queue_item_has_root_span
+                ):
+                    # if the currently processed span is a root span, OR the buffer
+                    # already had a root span inside, use a different timeout than
+                    # usual.
+                    if has_root_span:
+                        has_root_span_count += 1
+                        timestamp = now + self.span_buffer_root_timeout
```

Comment:
Is the intent to reset the timer every time a span is ingested, or to have the timer be based on when the root span was ingested?

Reply:
the timer is resetting everytime there is a new span. the timeout is a "timeout of inactivity", e.g. after 60 seconds of the buffer not being written to, we flush it